bokeh/models/sources.py

```diff
@@ -221,34 +221,23 @@ def columns(self, *columns):
         return ColumnsRef(source=self, columns=list(columns))
 
 
-    # def push_notebook(self):
-    #     """ Update date for a plot in the IPthon notebook in place.
-
-    #     This function can be be used to update data in plot data sources
-    #     in the IPython notebook, without having to use the Bokeh server.
-
-    #     Returns:
-    #         None
-
-    #     .. warning::
-    #         The current implementation leaks memory in the IPython notebook,
-    #         due to accumulating JS code. This function typically works well
-    #         with light UI interactions, but should not be used for continuously
-    #         updating data. See :bokeh-issue:`1732` for more details and to
-    #         track progress on potential fixes.
-
-    #     """
-    #     from IPython.core import display
-    #     from bokeh.protocol import serialize_json
-    #     id = self.ref['id']
-    #     model = self.ref['type']
-    #     json = serialize_json(self.vm_serialize())
-    #     js = """
-    #         var ds = Bokeh.Collections('{model}').get('{id}');
-    #         var data = {json};
-    #         ds.set(data);
-    #     """.format(model=model, id=id, json=json)
-    #     display.display_javascript(js, raw=True)
+    def push_notebook(self):
+        """ Update date for a plot in the IPthon notebook in place.
+
+        This function can be be used to update data in plot data sources
+        in the IPython notebook, without having to use the Bokeh server.
+
+        Returns:
+            None
+
+        """
+        import json
+        from ipykernel.comm import Comm
+
+        c = Comm(target_name='bokeh_notebook_comms',data={})
+
+        msg = json.dumps(dict(docid=self.document.id, source=self.to_json()))
+        c.send(msg)
```

Comment:
can we assume output_notebook here? if so we could have a notebook session id we establish in State and then use that to tell embed.coffee what to update. Rather than tunneling the id through Document, tunnel it through State.